src/pip/_internal/cli/progress_bars.py

```diff
@@ -55,6 +60,29 @@ def _rich_progress_bar(
             progress.update(task_id, advance=len(chunk))
 
 
+class _MachineReadableProgress:
+    def __init__(self, iterable: Iterable[bytes], size: Optional[int]) -> None:
+        self._iterable = iter(iterable)
+        self._size = size
+        self._progress = 0
+
+    def __iter__(self) -> Iterator[bytes]:
+        return self
+
+    def __next__(self) -> bytes:
+        chunk = next(self._iterable)
+        self._progress += len(chunk)
+        progress_info = {
+            "current": self._progress,
+            "total": self._size,
+        }
+        logger.info(
```

Comment:
I had that initially, but the way it was working was (for some reason) it was waiting until the download was finished to print everything at once... Which is definitely not what i want.

I can give it another try though

Reply:
Please do. `logger.info` isn't the right interface for this. Apart from anything else, it will make `-q --progress-bar=json` produce no progress bar, which is definitely not what you want.

The symptoms you describe suggest that either you weren't flushing output, or something in your subprocess setup left the stdio streams buffered in a way that's not a pip problem (but which suggests that relying on real-time progress output from pip might be trickier than you'd hoped in any case[^1]).

[^1]: For example, you might need to set the [`PYTHONUNBUFFERED`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUNBUFFERED) environment variable (or use the `-u` option) when invoking Python to run pip in the subprocess.